### PR TITLE
Cleanup of tests and crawler head

### PR DESCRIPTION
**Pull request by ricarkol**
Summary:
- Fix for the old bash functional tests in `tests/old/`
- New unit tests: [vms|hosts|containers]_crawler.py  …
- New functional tests: [vms|hosts|containers]_crawler.py  …
- crawler.py and crawlutils.py cleanup:
    + deleted crawlutils.py
    + added classes for vm, container and host crawling
    + added a Worker class to handle the scheduling of crawling and emits

### CODE REVIEW

**Comment by sahilsuneja1** [comment]
@ricarkol: some tests failing due to reference to crawlutils

**Comment by ricarkol** [comment]
Yeap, the tests side is not ready yet, that's mainly the work-in-progress.

File: crawler/containers_crawler.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        BaseCrawler.__init__(
+            self,
+            features=features,
+            plugin_places=plugin_places,
+            options=options)
+        self.containers_list = []
+        plugins_manager.reload_env_plugin(environment, plugin_places)
+        plugins_manager.reload_container_crawl_plugins(
+            features, plugin_places, options)
+        self.plugins = plugins_manager.get_container_crawl_plugins(features)
```
Comment (sahilsuneja1) [inline crawler/containers_crawler.py]: at some point (not in this PR) we should combine reload_xxx and get_xxx

File: crawler/containers_crawler.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        return frame
+
+    def crawl_containers(self):
+        """Returns a list generator of Frame objects"""
+        for container in self.containers_list:
+            yield self.crawl_container(container)
+
+    def crawl(self):
+        self.update_containers_list()
+        return self.crawl_containers()
```
Comment (sahilsuneja1) [inline crawler/containers_crawler.py]: very neat class!

File: crawler/crawler.conf [hunk 3]
```diff
@@ -2,21 +2,3 @@
     #enabled_plugins = os_container, cpu_container
 
 [ crawlers ]
```
Comment (sahilsuneja1) [inline crawler/crawler.conf]: Can you please bring this list back here, it doesn't affect which plugins get loaded, that is controlled by enabled_plugins in the above section

File: crawler/containers_logs_linker.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        dest='environment',
+        type=str,
+        default='cloudsight',
+        help='This speficies some environment specific behavior, like how '
+             'to name a container. The way to add a new behavior is by '
+             'implementing a plugin (see plugins/cloudsight_environment.py '
+             'as an example. Defaults to "cloudsight".',
+    )
+
+    misc.setup_logger('crawlutils', 'linker.log')
```
Comment (sahilsuneja1) [inline crawler/containers_logs_linker.py]: crawlutils still lives ;)

File: crawler/containers_logs_linker.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        for container in self.new:
+            container.link_logfiles()
+
+    def crawl(self):
+        self.update_containers_list()
+        self.link_containers()
+        return []
+
+
+if __name__ == '__main__':
```
Comment (sahilsuneja1) [inline crawler/containers_logs_linker.py]: why main here?

File: crawler/crawler.py [hunk 6]
```diff
@@ -325,41 +186,43 @@ def main():
...
-    params['crawlmode'] = args.crawlmode
-    params['format'] = args.format
+    misc.setup_logger('crawlutils', args.logfile)
+    misc.setup_logger('yapsy', 'yapsy.log')
 
     options = args.options
-    options['compress'] = (args.compress in ['true', 'True'])
     options['avoid_setns'] = args.avoid_setns
     options['mountpoint'] = args.mountpoint
-    options['vm_list'] = args.crawl_vm
```
Comment (sahilsuneja1) [inline crawler/crawler.py]: some of these would eventually come back right? e.g. vm_list?

File: crawler/crawlutils.py [hunk 7]
```diff
@@ -1,492 +0,0 @@
...
-        if (frequency < 0 or should_exit or snapshot_num == max_snapshots):
-            logger.info('Bye')
-            break
-
-        time_to_sleep, next_iteration_time = _get_next_iteration_time(
-            next_iteration_time, frequency, snapshot_time)
-        if time_to_sleep > 0:
-            time.sleep(time_to_sleep)
-
-        snapshot_num += 1
```
Comment (sahilsuneja1) [inline crawler/crawlutils.py]: did you smile or sighed?  ;)

File: crawler/plugins/disk_crawler.py [hunk 8]
```diff
@@ -9,13 +9,16 @@
...
+            pdiskusage = psutil.disk_usage(partition.mountpoint)
+            partitions.append((partition.mountpoint, DiskFeature(
+                partition.device,
+                100.0 - pdiskusage.percent,
+                partition.fstype,
+                partition.mountpoint,
+                partition.opts,
+                pdiskusage.total,
+            ), 'disk'))
+        except OSError:
```
Comment (sahilsuneja1) [inline crawler/plugins/disk_crawler.py]: what are we skipping here?

File: crawler/plugins_manager.py [hunk 9]
```diff
@@ -48,12 +51,11 @@ def _load_plugins(
         category_filter={},
         filter_func=lambda *arg: True,
         features=['os', 'cpu'],
+        plugin_places=['plugins'],
```
Comment (sahilsuneja1) [inline crawler/plugins_manager.py]: ah I get it, makes sens now that 'plugin_places' is no longer stored in 'options'

File: crawler/worker.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+            with Emitter(
+                    metadata=frame.metadata,
+                    snapshot_num=self.iter_count,
+                    **self.emitter_args
+            ) as emitter:
+                for (key, val, feature_type) in frame.data:
+                    emitter.emit(key, val, feature_type)
+
+        # just used for output purposes
+        self.iter_count += 1
```
Comment (sahilsuneja1) [inline crawler/worker.py]: again, neat class, I would reorder the functions  but that's just me.

File: crawler/containers_logs_linker.py (hunk 5 above)
Comment (ricarkol) [inline crawler/containers_logs_linker.py]: Maintaining logs wasn't fitting well with the main we already have. It's was like an extra weird feature, so decided to move it to its own main. The test for it (and how to use it) is `tests/functional/test_functional_logs_linker.py`.

File: crawler/containers_logs_linker.py (hunk 4 above)
Comment (ricarkol) [inline crawler/containers_logs_linker.py]: yup

File: crawler/crawler.py (hunk 6 above)
Comment (ricarkol) [inline crawler/crawler.py]: Added when adding the VirtualMachinesCrawler class.

File: crawler/crawlutils.py (hunk 7 above)
Comment (ricarkol) [inline crawler/crawlutils.py]: hehe, both

File: crawler/plugins/disk_crawler.py (hunk 8 above)
Comment (ricarkol) [inline crawler/plugins/disk_crawler.py]: If one of the partitions in psutil.disk_partitions disappears while we are in the loop, an OSError is triggered.

**Comment by sahilsuneja1** [comment]
Great! Merging.

**Pull request merged**